lua/neotest-go/init.lua

```diff
@@ -82,60 +89,189 @@ local function get_go_package_name(_)
   return vim.startswith('package', line) and vim.split(line, ' ')[2] or ''
 end
 
+--- gets the root directory of the go project
+---@param start_file string
+---@return string?
+local function get_go_root(start_file)
+  return lib.files.match_root_pattern('go.mod')(start_file)
+end
+
+--- gets the go module name
+---@param go_root string
+---@return string?
+local function get_go_module_name(go_root)
+  local gomod_file = go_root .. '/go.mod'
+  local gomod_success, gomodule = pcall(lib.files.read_lines, gomod_file)
+  if not gomod_success then
+    logger.error("neotest-go: couldn't read go.mod file: " .. gomodule)
+    return
+  end
+  local line = gomodule[1]
+  local module = string.match(line, 'module (.+)')
+  return module
+end
+
 local function get_experimental_opts()
   return {
-    test_table = false
+    test_table = false,
   }
 end
 
 local get_args = function()
   return {}
 end
 
+--- Converts from a given go package and the "/" seperated testname to a
+--- format "package::test::subtest".
+--- Returns the test in this format as well as the testname of the parent test (if present)
+---@param package string
+---@param test string
+---@return string, string?
+local function normalize_testname(package, test)
+  -- sub-tests are structured as 'TestMainTest/subtest_clause'
+  local parts = vim.split(test, '/')
+  local is_subtest = #parts > 1
+  local parenttest = is_subtest and (package .. '::' .. parts[1]) or nil
+  return package .. '::' .. table.concat(parts, '::'), parenttest
+end
+
+--- Converts from a given neotest id and go_root / go_module to format
+--- "package::test::subtest"
+---@param id string
+---@param go_root string
+---@param go_module string
+---@return string
+local function normalize_id(id, go_root, go_module)
+  local normalized_id, _ = id:gsub(go_root, go_module):gsub('/%w*_test.go', '')
+  return normalized_id
+end
+
+--- Extracts the file name from a neotest id
+---@param id string
+---@return string
+local function get_filename_from_id(id)
+  local filename = string.match(id, '/(%w*_test.go)::')
+  return filename
+end
+
+--- Extracts testfile and linenumber of go test output in format
+--- "    main_test.go:12: Some error message\n"
+---@param line string
+---@return string?, number?
+local function get_testfileinfo(line)
+  if line then
+    local file, linenumber = string.match(line, testfile_pattern)
+    return file, tonumber(linenumber)
+  end
+  return nil, nil
+end
+
+local function is_error(lines)
+  for _, line in ipairs(lines) do
+    line = string.lower(line)
+    for _, pattern in ipairs(error_pattern) do
+      if string.match(line, string.lower(pattern)) then
```

Comment:
Similar to the above, where possible I'd say to just use the method form of string functions on the string itself
```suggestion
      if line:match(pattern:lower()) then
```